File: weile_ddz/cards.py
"""54 张牌编码：S/H/C/D + 点数；SJ=小王，BJ=大王。例：S_A、H_2。"""

from __future__ import annotations

from dataclasses import dataclass

SUITS = (
    ("S", "♠", "黑桃", "black"),
    ("H", "♥", "红桃", "red"),
    ("C", "♣", "梅花", "black"),
    ("D", "♦", "方片", "red"),
)

RANK_LABELS = ("3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A", "2")
RANK_VALUE = {label: index + 3 for index, label in enumerate(RANK_LABELS)}
RANK_VALUE["SJ"] = 16
RANK_VALUE["BJ"] = 17

SEQUENCE_MIN = 3
SEQUENCE_MAX = 14  # A
SEQUENCE_RANKS = {label for label, value in RANK_VALUE.items() if 3 <= value <= 14}

VALUE_LABEL = {value: label for label, value in RANK_VALUE.items()}
VALUE_LABEL[16] = "小王"
VALUE_LABEL[17] = "大王"


@dataclass(frozen=True)
class Card:
    id: str
    rank: str
    value: int
    suit: str | None
    symbol: str
    label: str
    name: str
    color: str
    is_joker: bool

    def display(self) -> str:
        if self.is_joker:
            return self.label
        return f"{self.symbol}{self.label}"


def _build_deck() -> list[Card]:
    cards: list[Card] = []
    for rank in RANK_LABELS:
        for suit, symbol, suit_name, color in SUITS:
            card_id = f"{suit}_{rank}"
            cards.append(
                Card(
                    id=card_id,
                    rank=rank,
                    value=RANK_VALUE[rank],
                    suit=suit,
                    symbol=symbol,
                    label=rank,
                    name=f"{suit_name}{rank}",
                    color=color,
                    is_joker=False,
                )
            )
    cards.append(
        Card("SJ", "SJ", 16, None, "☆", "小王", "小王", "joker-s", True)
    )
    cards.append(
        Card("BJ", "BJ", 17, None, "★", "大王", "大王", "joker-b", True)
    )
    return cards


DECK = _build_deck()
CARD_BY_ID = {card.id: card for card in DECK}
ALL_IDS = [card.id for card in DECK]


def cards_by_ids(ids: list[str]) -> list[Card]:
    return [CARD_BY_ID[card_id] for card_id in ids]


def sort_cards(cards: list[Card]) -> list[Card]:
    return sorted(cards, key=lambda card: (card.value, card.suit or "Z"))


def format_cards(cards: list[Card]) -> str:
    return " ".join(card.display() for card in sort_cards(cards))


def format_ranks(cards: list[Card]) -> str:
    return " ".join(card.label for card in sort_cards(cards))


def rank_counts(cards: list[Card]) -> dict[int, int]:
    counts: dict[int, int] = {}
    for card in cards:
        counts[card.value] = counts.get(card.value, 0) + 1
    return counts


def value_label(value: int) -> str:
    return VALUE_LABEL.get(value, str(value))

File: games/weile-doudizhu/js/cards.js
/** 微乐经典斗地主：一副 54 张（4 花色 × 13 点 + 小王 + 大王）。 */

export const SUITS = [
  { id: "S", symbol: "♠", name: "黑桃", color: "black" },
  { id: "H", symbol: "♥", name: "红桃", color: "red" },
  { id: "C", symbol: "♣", name: "梅花", color: "black" },
  { id: "D", symbol: "♦", name: "方片", color: "red" },
];

/** 点数从 3 到大王，数值越大越大。2=15，小王=16，大王=17。 */
export const RANK_VALUES = [3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17];

export const RANK_LABEL = {
  3: "3",
  4: "4",
  5: "5",
  6: "6",
  7: "7",
  8: "8",
  9: "9",
  10: "10",
  11: "J",
  12: "Q",
  13: "K",
  14: "A",
  15: "2",
  16: "小王",
  17: "大王",
};

/** 可参与顺子 / 连对 / 飞机主体的点数（3–A）。2 与王不行。 */
export const SEQUENCE_MAX_RANK = 14;
export const SEQUENCE_MIN_RANK = 3;

export function rankLabel(rank) {
  return RANK_LABEL[rank] ?? String(rank);
}

export function createDeck() {
  const cards = [];
  for (const rank of RANK_VALUES) {
    if (rank >= 16) continue;
    for (const suit of SUITS) {
      cards.push({
        id: `${suit.id}${rank}`,
        rank,
        suit: suit.id,
        label: RANK_LABEL[rank],
        symbol: suit.symbol,
        color: suit.color,
        isJoker: false,
        name: `${suit.name}${RANK_LABEL[rank]}`,
      });
    }
  }
  cards.push({
    id: "JOX",
    rank: 16,
    suit: "J",
    label: "小王",
    symbol: "☆",
    color: "joker-s",
    isJoker: true,
    name: "小王",
  });
  cards.push({
    id: "JOY",
    rank: 17,
    suit: "J",
    label: "大王",
    symbol: "★",
    color: "joker-b",
    isJoker: true,
    name: "大王",
  });
  return cards;
}

export const DECK = createDeck();
export const CARD_BY_ID = Object.fromEntries(DECK.map((card) => [card.id, card]));

export function cardsByIds(ids) {
  return ids.map((id) => {
    const card = CARD_BY_ID[id];
    if (!card) throw new Error(`未知牌 id: ${id}`);
    return card;
  });
}

export function sortCards(cards) {
  return [...cards].sort((a, b) => a.rank - b.rank || a.suit.localeCompare(b.suit));
}

export function formatCards(cards) {
  return sortCards(cards)
    .map((card) => (card.isJoker ? card.label : `${card.symbol}${card.label}`))
    .join(" ");
}

export function formatRanks(cards) {
  return sortCards(cards)
    .map((card) => card.label)
    .join(" ");
}

export function rankCountsFromCards(cards) {
  const counts = new Map();
  for (const card of cards) {
    counts.set(card.rank, (counts.get(card.rank) || 0) + 1);
  }
  return counts;
}

export function cardsOfRank(cards, rank) {
  return cards.filter((card) => card.rank === rank);
}

export function uniqueIds(ids) {
  return [...new Set(ids)];
}
